### PR TITLE
Move legacy TH functions(sinh,cosh) to TensorIterator + Vec256

### PR DESCRIPTION
**Pull request by xta0**
This is a follow up on Jame's PR: https://github.com/pytorch/pytorch/pull/19041. The idea is to replace the legacy `sinh` / `cosh` ops that are being dispatched to TH with the operations defined in `Vec256` for better performance.

benchmark(from Jame's script):

```python
import torch, time
ops = ['sinh', 'cosh']
x = torch.rand(1024, 1024)
NITER = 10000

print('op', 'time per iter (ms)', 'gops/s', 'GB/s', sep='\t')
for op in ops:
    s = time.time()
    for i in range(NITER):
        getattr(x, op)()
    elapsed_sec = ((time.time() - s) / NITER)
    print(op, elapsed_sec * 1000, (1024*1024/elapsed_sec)/1e9, (1024*1024*4*2) / elapsed_sec / 1e9, sep='\t')
```
code on master:

```
op	time per iter (ms)	gops/s	GB/s
sinh	3.37614369392395	0.3105839369002935	2.484671495202348
cosh	3.480502033233643	0.3012714803748572	2.4101718429988574
```
after change (on Macbook pro 2018):

```
op	time per iter (ms)	gops/s	GB/s
sinh	0.8956503868103027	1.1707425301677301	9.365940241341841
cosh	0.9392147302627564	1.1164390487217428	8.931512389773943
```

### CODE REVIEW

File: aten/src/ATen/native/UnaryOps.cpp [hunk 1]
```diff
@@ -181,7 +181,7 @@ IMPLEMENT_UNARY_OP_VEC(asin)
 IMPLEMENT_UNARY_OP_VEC(atan)
 IMPLEMENT_UNARY_OP_VEC(ceil)
 IMPLEMENT_UNARY_OP_VEC(cos)
-IMPLEMENT_UNARY_OP_TH(cosh)
+IMPLEMENT_UNARY_OP_VEC(cosh)
```
Comment (ljk53) [inline aten/src/ATen/native/UnaryOps.cpp]: Can we remove the IMPLEMENT_UNARY_OP_TH macro defined above?

**Comment by ljk53** [comment]
And did you try removing TH cpu backend for _th_cosh & _th_sinh from aten/src/ATen/Declarations.cwrap?

File: aten/src/ATen/native/UnaryOps.cpp (hunk 1 above)
Comment (xta0) [reply aten/src/ATen/native/UnaryOps.cpp]: Sounds good.

**Comment by cpuhrsch** [comment]
Please notice [this comment](https://github.com/pytorch/pytorch/blob/43f94077d8bed5965ff47fd37f9611282767e783/aten/src/ATen/cpu/vml.h#L165) and also [this comment](https://github.com/pytorch/pytorch/blob/43f94077d8bed5965ff47fd37f9611282767e783/aten/src/ATen/cpu/vml.h#L37)

cc @VitalyFedyunin

**Comment by cpuhrsch** [comment]
glibc's math library doesn't vectorize sinh and cosh and the float specializations of vec256's sinh and cosh doesn't use anything better than that.

**Comment by cpuhrsch** [comment]
The [sleef](https://sleef.org/purec.xhtml) sinh operations are limit in ULP1 range, whereas [glibc](https://www.gnu.org/software/libc/manual/html_node/Errors-in-Math-Functions.html) guarantees ULP2 for the full range. Looking at the function this sort of doesn't make any sense in the tails (since the function grows so large) and yet we want to still produce values that are standardized on something. We also don't want to produce 0s, which is what SLEEF would do.

Now, [VML](https://software.intel.com/en-us/mkl-developer-reference-c-v-sinh#EEA1269B-58A8-40A2-901C-20346EF68808) is compliant with glibc as far as I know, but there are issues with clang as mentioned. sinh and cosh were [disabled](https://github.com/pytorch/pytorch/pull/8488) as part of this PR. It's true that I should have added more detail, but we have yet to do a full investigation into why this fails for clang or at least disable it for clang.

I'd say the better strategy is to
1. Reenable VML for sinh and cosh via [the performant macro used for the other unary functions](https://github.com/pytorch/pytorch/pull/21115/files#diff-5c451ccc1198b8c48f2b768b47e87a8aL174) and within the [vml header](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cpu/vml.h)
2. Closely monitor clang's behavior and investigate any failures.

But the current PR likely mostly takes advantage of the improved memory read patterns. Take care that the transition penalty bug isn't happening and see how other Unary functions deal with it.

**Comment by xta0** [comment]
> And did you try removing TH cpu backend for _th_cosh & _th_sinh from aten/src/ATen/Declarations.cwrap?

Yes, I've removed the cpu backend for both of them.

**Comment by xta0** [comment]
> glibc's math library doesn't vectorize sinh and cosh and the float specializations of vec256's sinh and cosh doesn't use anything better than that.

Yea, I switched the method from `unary_kernel_vec` to `unary_kernel`.

**Comment by fmassa** [comment]
While you are st it, can you also do the same thing for CUDA? It shouldn't be much more work, and makes everything simpler

File: aten/src/ATen/native/UnaryOps.cpp [hunk 2]
```diff
@@ -181,7 +165,7 @@ IMPLEMENT_UNARY_OP_VEC(asin)
...
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/UnaryOps.cpp]: Can we comment here, that kernels are not using Vec256 (because _VEC is confusing )

**Review by VitalyFedyunin — APPROVED** [review]
Code looks good. PR would become perfect if you remove dead TH operators code.
